File: android/src/main/java/com/apollohg/editor/EditorTheme.kt
package com.apollohg.editor

import android.graphics.Color
import android.graphics.Typeface
import org.json.JSONObject

data class EditorTextStyle(
    val fontFamily: String? = null,
    val fontSize: Float? = null,
    val fontWeight: String? = null,
    val fontStyle: String? = null,
    val color: Int? = null,
    val lineHeight: Float? = null,
    val spacingAfter: Float? = null
) {
    companion object {
        fun fromJson(json: JSONObject?): EditorTextStyle? {
            json ?: return null
            return EditorTextStyle(
                fontFamily = json.optNullableString("fontFamily"),
                fontSize = json.optNullableFloat("fontSize"),
                fontWeight = json.optNullableString("fontWeight"),
                fontStyle = json.optNullableString("fontStyle"),
                color = parseColor(json.optNullableString("color")),
                lineHeight = json.optNullableFloat("lineHeight"),
                spacingAfter = json.optNullableFloat("spacingAfter")
            )
        }
    }

    fun mergedWith(other: EditorTextStyle?): EditorTextStyle {
        other ?: return this
        return copy(
            fontFamily = other.fontFamily ?: fontFamily,
            fontSize = other.fontSize ?: fontSize,
            fontWeight = other.fontWeight ?: fontWeight,
            fontStyle = other.fontStyle ?: fontStyle,
            color = other.color ?: color,
            lineHeight = other.lineHeight ?: lineHeight,
            spacingAfter = other.spacingAfter ?: spacingAfter
        )
    }

    fun typefaceStyle(): Int {
        val bold = fontWeight == "bold" || fontWeight?.toIntOrNull()?.let { it >= 600 } == true
        val italic = fontStyle == "italic"
        return when {
            bold && italic -> Typeface.BOLD_ITALIC
            bold -> Typeface.BOLD
            italic -> Typeface.ITALIC
            else -> Typeface.NORMAL
        }
    }
}

data class EditorListTheme(
    val indent: Float? = null,
    val baseIndentMultiplier: Float? = null,
    val itemSpacing: Float? = null,
    val markerColor: Int? = null,
    val markerScale: Float? = null
) {
    companion object {
        fun fromJson(json: JSONObject?): EditorListTheme? {
            json ?: return null
            return EditorListTheme(
                indent = json.optNullableFloat("indent"),
                baseIndentMultiplier = json.optNullableFloat("baseIndentMultiplier"),
                itemSpacing = json.optNullableFloat("itemSpacing"),
                markerColor = parseColor(json.optNullableString("markerColor")),
                markerScale = json.optNullableFloat("markerScale")
            )
        }
    }
}

data class EditorHorizontalRuleTheme(
    val color: Int? = null,
    val thickness: Float? = null,
    val verticalMargin: Float? = null
) {
    companion object {
        fun fromJson(json: JSONObject?): EditorHorizontalRuleTheme? {
            json ?: return null
            return EditorHorizontalRuleTheme(
                color = parseColor(json.optNullableString("color")),
                thickness = json.optNullableFloat("thickness"),
                verticalMargin = json.optNullableFloat("verticalMargin")
            )
        }
    }
}

data class EditorBlockquoteTheme(
    val text: EditorTextStyle? = null,
    val indent: Float? = null,
    val borderColor: Int? = null,
    val borderWidth: Float? = null,
    val markerGap: Float? = null
) {
    companion object {
        fun fromJson(json: JSONObject?): EditorBlockquoteTheme? {
            json ?: return null
            return EditorBlockquoteTheme(
                text = EditorTextStyle.fromJson(json.optJSONObject("text")),
                indent = json.optNullableFloat("indent"),
                borderColor = parseColor(json.optNullableString("borderColor")),
                borderWidth = json.optNullableFloat("borderWidth"),
                markerGap = json.optNullableFloat("markerGap")
            )
        }
    }
}

data class EditorLinkTheme(
    val fontFamily: String? = null,
    val fontSize: Float? = null,
    val fontWeight: String? = null,
    val fontStyle: String? = null,
    val color: Int? = null,
    val backgroundColor: Int? = null,
    val underline: Boolean? = null
) {
    companion object {
        fun fromJson(json: JSONObject?): EditorLinkTheme? {
            json ?: return null
            return EditorLinkTheme(
                fontFamily = json.optNullableString("fontFamily"),
                fontSize = json.optNullableFloat("fontSize"),
                fontWeight = json.optNullableString("fontWeight"),
                fontStyle = json.optNullableString("fontStyle"),
                color = parseColor(json.optNullableString("color")),
                backgroundColor = parseColor(json.optNullableString("backgroundColor")),
                underline = if (json.has("underline")) json.optBoolean("underline") else null
            )
        }
    }

    fun asTextStyle(): EditorTextStyle =
        EditorTextStyle(
            fontFamily = fontFamily,
            fontSize = fontSize,
            fontWeight = fontWeight,
            fontStyle = fontStyle,
            color = color
        )
}

data class EditorMentionTheme(
    val textColor: Int? = null,
    val backgroundColor: Int? = null,
    val borderColor: Int? = null,
    val borderWidth: Float? = null,
    val borderRadius: Float? = null,
    val fontWeight: String? = null,
    val popoverBackgroundColor: Int? = null,
    val popoverBorderColor: Int? = null,
    val popoverBorderWidth: Float? = null,
    val popoverBorderRadius: Float? = null,
    val popoverShadowColor: Int? = null,
    val optionTextColor: Int? = null,
    val optionSecondaryTextColor: Int? = null,
    val optionHighlightedBackgroundColor: Int? = null,
    val optionHighlightedTextColor: Int? = null
) {
    fun mergedWith(other: EditorMentionTheme?): EditorMentionTheme {
        other ?: return this
        return copy(
            textColor = other.textColor ?: textColor,
            backgroundColor = other.backgroundColor ?: backgroundColor,
            borderColor = other.borderColor ?: borderColor,
            borderWidth = other.borderWidth ?: borderWidth,
            borderRadius = other.borderRadius ?: borderRadius,
            fontWeight = other.fontWeight ?: fontWeight,
            popoverBackgroundColor = other.popoverBackgroundColor ?: popoverBackgroundColor,
            popoverBorderColor = other.popoverBorderColor ?: popoverBorderColor,
            popoverBorderWidth = other.popoverBorderWidth ?: popoverBorderWidth,
            popoverBorderRadius = other.popoverBorderRadius ?: popoverBorderRadius,
            popoverShadowColor = other.popoverShadowColor ?: popoverShadowColor,
            optionTextColor = other.optionTextColor ?: optionTextColor,
            optionSecondaryTextColor = other.optionSecondaryTextColor ?: optionSecondaryTextColor,
            optionHighlightedBackgroundColor =
                other.optionHighlightedBackgroundColor ?: optionHighlightedBackgroundColor,
            optionHighlightedTextColor =
                other.optionHighlightedTextColor ?: optionHighlightedTextColor
        )
    }

    companion object {
        fun fromJson(json: JSONObject?): EditorMentionTheme? {
            json ?: return null
            return EditorMentionTheme(
                textColor = parseColor(json.optNullableString("textColor")),
                backgroundColor = parseColor(json.optNullableString("backgroundColor")),
                borderColor = parseColor(json.optNullableString("borderColor")),
                borderWidth = json.optNullableFloat("borderWidth"),
                borderRadius = json.optNullableFloat("borderRadius"),
                fontWeight = json.optNullableString("fontWeight"),
                popoverBackgroundColor = parseColor(json.optNullableString("popoverBackgroundColor")),
                popoverBorderColor = parseColor(json.optNullableString("popoverBorderColor")),
                popoverBorderWidth = json.optNullableFloat("popoverBorderWidth"),
                popoverBorderRadius = json.optNullableFloat("popoverBorderRadius"),
                popoverShadowColor = parseColor(json.optNullableString("popoverShadowColor")),
                optionTextColor = parseColor(json.optNullableString("optionTextColor")),
                optionSecondaryTextColor = parseColor(json.optNullableString("optionSecondaryTextColor")),
                optionHighlightedBackgroundColor = parseColor(json.optNullableString("optionHighlightedBackgroundColor")),
                optionHighlightedTextColor = parseColor(json.optNullableString("optionHighlightedTextColor"))
            )
        }
    }
}

enum class EditorToolbarAppearance {
    CUSTOM,
    NATIVE;

    companion object {
        fun fromRaw(raw: String?): EditorToolbarAppearance? =
            when (raw?.trim()?.lowercase()) {
                "custom" -> CUSTOM
                "native" -> NATIVE
                else -> null
            }
    }
}

data class EditorToolbarTheme(
    val appearance: EditorToolbarAppearance? = null,
    val backgroundColor: Int? = null,
    val borderColor: Int? = null,
    val borderWidth: Float? = null,
    val borderRadius: Float? = null,
    val marginTop: Float? = null,
    val showTopBorder: Boolean? = null,
    val keyboardOffset: Float? = null,
    val horizontalInset: Float? = null,
    val separatorColor: Int? = null,
    val buttonColor: Int? = null,
    val buttonActiveColor: Int? = null,
    val buttonDisabledColor: Int? = null,
    val buttonActiveBackgroundColor: Int? = null,
    val buttonBorderRadius: Float? = null
) {
    fun resolvedKeyboardOffset(): Float = keyboardOffset ?: if (appearance == EditorToolbarAppearance.NATIVE) 8f else 0f

    fun resolvedHorizontalInset(): Float = horizontalInset ?: if (appearance == EditorToolbarAppearance.NATIVE) 0f else 0f

    fun resolvedBorderRadius(): Float = if (appearance == EditorToolbarAppearance.NATIVE) 32f else (borderRadius ?: 0f)

    fun resolvedBorderWidth(): Float = borderWidth ?: if (appearance == EditorToolbarAppearance.NATIVE) 0f else 1f

    fun resolvedButtonBorderRadius(): Float = if (appearance == EditorToolbarAppearance.NATIVE) 20f else (buttonBorderRadius ?: 6f)

    companion object {
        fun fromJson(json: JSONObject?): EditorToolbarTheme? {
            json ?: return null
            return EditorToolbarTheme(
                appearance = EditorToolbarAppearance.fromRaw(json.optNullableString("appearance")),
                backgroundColor = parseColor(json.optNullableString("backgroundColor")),
                borderColor = parseColor(json.optNullableString("borderColor")),
                borderWidth = json.optNullableFloat("borderWidth"),
                borderRadius = json.optNullableFloat("borderRadius"),
                marginTop = json.optNullableFloat("marginTop"),
                showTopBorder = if (json.has("showTopBorder")) json.optBoolean("showTopBorder") else null,
                keyboardOffset = json.optNullableFloat("keyboardOffset"),
                horizontalInset = json.optNullableFloat("horizontalInset"),
                separatorColor = parseColor(json.optNullableString("separatorColor")),
                buttonColor = parseColor(json.optNullableString("buttonColor")),
                buttonActiveColor = parseColor(json.optNullableString("buttonActiveColor")),
                buttonDisabledColor = parseColor(json.optNullableString("buttonDisabledColor")),
                buttonActiveBackgroundColor = parseColor(json.optNullableString("buttonActiveBackgroundColor")),
                buttonBorderRadius = json.optNullableFloat("buttonBorderRadius")
            )
        }
    }
}

data class EditorContentInsets(
    val top: Float? = null,
    val right: Float? = null,
    val bottom: Float? = null,
    val left: Float? = null
) {
    companion object {
        fun fromJson(json: JSONObject?): EditorContentInsets? {
            json ?: return null
            return EditorContentInsets(
                top = json.optNullableFloat("top"),
                right = json.optNullableFloat("right"),
                bottom = json.optNullableFloat("bottom"),
                left = json.optNullableFloat("left")
            )
        }
    }
}

data class EditorTheme(
    val text: EditorTextStyle? = null,
    val paragraph: EditorTextStyle? = null,
    val blockquote: EditorBlockquoteTheme? = null,
    val headings: Map<String, EditorTextStyle> = emptyMap(),
    val list: EditorListTheme? = null,
    val horizontalRule: EditorHorizontalRuleTheme? = null,
    val mentions: EditorMentionTheme? = null,
    val links: EditorLinkTheme? = null,
    val toolbar: EditorToolbarTheme? = null,
    val backgroundColor: Int? = null,
    val borderRadius: Float? = null,
    val contentInsets: EditorContentInsets? = null
) {
    companion object {
        fun fromJson(json: String?): EditorTheme? {
            if (json.isNullOrBlank()) return null
            val root = try {
                JSONObject(json)
            } catch (_: Exception) {
                return null
            }

            val headings = mutableMapOf<String, EditorTextStyle>()
            for (level in listOf("h1", "h2", "h3", "h4", "h5", "h6")) {
                val style = EditorTextStyle.fromJson(root.optJSONObject("headings")?.optJSONObject(level))
                if (style != null) {
                    headings[level] = style
                }
            }

            return EditorTheme(
                text = EditorTextStyle.fromJson(root.optJSONObject("text")),
                paragraph = EditorTextStyle.fromJson(root.optJSONObject("paragraph")),
                blockquote = EditorBlockquoteTheme.fromJson(root.optJSONObject("blockquote")),
                headings = headings,
                list = EditorListTheme.fromJson(root.optJSONObject("list")),
                horizontalRule = EditorHorizontalRuleTheme.fromJson(root.optJSONObject("horizontalRule")),
                mentions = EditorMentionTheme.fromJson(root.optJSONObject("mentions")),
                links = EditorLinkTheme.fromJson(root.optJSONObject("links")),
                toolbar = EditorToolbarTheme.fromJson(root.optJSONObject("toolbar")),
                backgroundColor = parseColor(root.optNullableString("backgroundColor")),
                borderRadius = root.optNullableFloat("borderRadius"),
                contentInsets = EditorContentInsets.fromJson(root.optJSONObject("contentInsets"))
            )
        }
    }

    fun effectiveTextStyle(nodeType: String, inBlockquote: Boolean = false): EditorTextStyle {
        var style = text ?: EditorTextStyle()
        style = style.mergedWith(if (inBlockquote) blockquote?.text else null)
        if (nodeType == "paragraph") {
            style = style.mergedWith(paragraph)
            if (paragraph?.lineHeight == null) {
                style = style.copy(lineHeight = null)
            }
        }
        style = style.mergedWith(headings[nodeType])
        return style
    }
}

private fun parseColor(raw: String?): Int? {
    val value = raw?.trim()?.lowercase() ?: return null
    if (value.isEmpty()) return null

    parseCssHexColor(value)?.let { return it }

    try {
        return Color.parseColor(value)
    } catch (_: IllegalArgumentException) {
        // Fall through to rgb()/rgba() parsing.
    }

    return when {
        value.startsWith("rgb(") && value.endsWith(")") -> {
            val parts = value.removePrefix("rgb(").removeSuffix(")")
                .split(',')
                .map { it.trim() }
            if (parts.size != 3) return null
            val red = parts[0].toDoubleOrNull() ?: return null
            val green = parts[1].toDoubleOrNull() ?: return null
            val blue = parts[2].toDoubleOrNull() ?: return null
            Color.argb(255, red.toInt(), green.toInt(), blue.toInt())
        }
        value.startsWith("rgba(") && value.endsWith(")") -> {
            val parts = value.removePrefix("rgba(").removeSuffix(")")
                .split(',')
                .map { it.trim() }
            if (parts.size != 4) return null
            val red = parts[0].toDoubleOrNull() ?: return null
            val green = parts[1].toDoubleOrNull() ?: return null
            val blue = parts[2].toDoubleOrNull() ?: return null
            val alpha = parts[3].toDoubleOrNull() ?: return null
            Color.argb((alpha * 255f).toInt(), red.toInt(), green.toInt(), blue.toInt())
        }
        else -> null
    }
}

private fun parseCssHexColor(value: String): Int? {
    if (!value.startsWith("#")) return null
    val hex = value.removePrefix("#")

    return when (hex.length) {
        3 -> {
            val red = "${hex[0]}${hex[0]}".toIntOrNull(16) ?: return null
            val green = "${hex[1]}${hex[1]}".toIntOrNull(16) ?: return null
            val blue = "${hex[2]}${hex[2]}".toIntOrNull(16) ?: return null
            Color.argb(255, red, green, blue)
        }
        4 -> {
            val red = "${hex[0]}${hex[0]}".toIntOrNull(16) ?: return null
            val green = "${hex[1]}${hex[1]}".toIntOrNull(16) ?: return null
            val blue = "${hex[2]}${hex[2]}".toIntOrNull(16) ?: return null
            val alpha = "${hex[3]}${hex[3]}".toIntOrNull(16) ?: return null
            Color.argb(alpha, red, green, blue)
        }
        6 -> {
            val red = hex.substring(0, 2).toIntOrNull(16) ?: return null
            val green = hex.substring(2, 4).toIntOrNull(16) ?: return null
            val blue = hex.substring(4, 6).toIntOrNull(16) ?: return null
            Color.argb(255, red, green, blue)
        }
        8 -> {
            val red = hex.substring(0, 2).toIntOrNull(16) ?: return null
            val green = hex.substring(2, 4).toIntOrNull(16) ?: return null
            val blue = hex.substring(4, 6).toIntOrNull(16) ?: return null
            val alpha = hex.substring(6, 8).toIntOrNull(16) ?: return null
            Color.argb(alpha, red, green, blue)
        }
        else -> null
    }
}

private fun JSONObject.optNullableString(key: String): String? {
    if (!has(key) || isNull(key)) return null
    return optString(key).takeUnless { it == "null" }
}

private fun JSONObject.optNullableFloat(key: String): Float? {
    if (!has(key) || isNull(key)) return null
    return optDouble(key).takeIf { !it.isNaN() }?.toFloat()
}

File: ios/EditorTheme.swift
import UIKit

struct EditorTextStyle {
    var fontFamily: String?
    var fontSize: CGFloat?
    var fontWeight: String?
    var fontStyle: String?
    var color: UIColor?
    var lineHeight: CGFloat?
    var spacingAfter: CGFloat?

    init(
        fontFamily: String? = nil,
        fontSize: CGFloat? = nil,
        fontWeight: String? = nil,
        fontStyle: String? = nil,
        color: UIColor? = nil,
        lineHeight: CGFloat? = nil,
        spacingAfter: CGFloat? = nil
    ) {
        self.fontFamily = fontFamily
        self.fontSize = fontSize
        self.fontWeight = fontWeight
        self.fontStyle = fontStyle
        self.color = color
        self.lineHeight = lineHeight
        self.spacingAfter = spacingAfter
    }

    init(dictionary: [String: Any]) {
        fontFamily = dictionary["fontFamily"] as? String
        fontSize = EditorTheme.cgFloat(dictionary["fontSize"])
        fontWeight = dictionary["fontWeight"] as? String
        fontStyle = dictionary["fontStyle"] as? String
        color = EditorTheme.color(from: dictionary["color"])
        lineHeight = EditorTheme.cgFloat(dictionary["lineHeight"])
        spacingAfter = EditorTheme.cgFloat(dictionary["spacingAfter"])
    }

    func merged(with override: EditorTextStyle?) -> EditorTextStyle {
        guard let override else { return self }
        return EditorTextStyle(
            fontFamily: override.fontFamily ?? fontFamily,
            fontSize: override.fontSize ?? fontSize,
            fontWeight: override.fontWeight ?? fontWeight,
            fontStyle: override.fontStyle ?? fontStyle,
            color: override.color ?? color,
            lineHeight: override.lineHeight ?? lineHeight,
            spacingAfter: override.spacingAfter ?? spacingAfter
        )
    }

    func resolvedFont(fallback: UIFont) -> UIFont {
        let size = fontSize ?? fallback.pointSize
        var font = fallback.withSize(size)

        if let fontFamily,
           let familyFont = UIFont(name: fontFamily, size: size) {
            font = familyFont
        } else if let fontWeight {
            font = UIFont.systemFont(ofSize: size, weight: EditorTheme.fontWeight(from: fontWeight))
        }

        var traits = font.fontDescriptor.symbolicTraits
        if EditorTheme.shouldApplyBoldTrait(fontWeight) {
            traits.insert(.traitBold)
        }
        if fontStyle == "italic" {
            traits.insert(.traitItalic)
        }

        if traits != font.fontDescriptor.symbolicTraits,
           let descriptor = font.fontDescriptor.withSymbolicTraits(traits) {
            font = UIFont(descriptor: descriptor, size: size)
        }

        return font
    }
}

struct EditorListTheme {
    var indent: CGFloat?
    var baseIndentMultiplier: CGFloat?
    var itemSpacing: CGFloat?
    var markerColor: UIColor?
    var markerScale: CGFloat?

    init(dictionary: [String: Any]) {
        indent = EditorTheme.cgFloat(dictionary["indent"])
        baseIndentMultiplier = EditorTheme.cgFloat(dictionary["baseIndentMultiplier"])
        itemSpacing = EditorTheme.cgFloat(dictionary["itemSpacing"])
        markerColor = EditorTheme.color(from: dictionary["markerColor"])
        markerScale = EditorTheme.cgFloat(dictionary["markerScale"])
    }
}

struct EditorHorizontalRuleTheme {
    var color: UIColor?
    var thickness: CGFloat?
    var verticalMargin: CGFloat?

    init(dictionary: [String: Any]) {
        color = EditorTheme.color(from: dictionary["color"])
        thickness = EditorTheme.cgFloat(dictionary["thickness"])
        verticalMargin = EditorTheme.cgFloat(dictionary["verticalMargin"])
    }
}

struct EditorBlockquoteTheme {
    var text: EditorTextStyle?
    var indent: CGFloat?
    var borderColor: UIColor?
    var borderWidth: CGFloat?
    var markerGap: CGFloat?

    init(dictionary: [String: Any]) {
        if let text = dictionary["text"] as? [String: Any] {
            self.text = EditorTextStyle(dictionary: text)
        }
        indent = EditorTheme.cgFloat(dictionary["indent"])
        borderColor = EditorTheme.color(from: dictionary["borderColor"])
        borderWidth = EditorTheme.cgFloat(dictionary["borderWidth"])
        markerGap = EditorTheme.cgFloat(dictionary["markerGap"])
    }
}

struct EditorLinkTheme {
    var fontFamily: String?
    var fontSize: CGFloat?
    var fontWeight: String?
    var fontStyle: String?
    var color: UIColor?
    var backgroundColor: UIColor?
    var underline: Bool?

    init(dictionary: [String: Any]) {
        fontFamily = dictionary["fontFamily"] as? String
        fontSize = EditorTheme.cgFloat(dictionary["fontSize"])
        fontWeight = dictionary["fontWeight"] as? String
        fontStyle = dictionary["fontStyle"] as? String
        color = EditorTheme.color(from: dictionary["color"])
        backgroundColor = EditorTheme.color(from: dictionary["backgroundColor"])
        underline = dictionary["underline"] as? Bool
    }

    func resolvedFont(fallback: UIFont) -> UIFont {
        EditorTextStyle(
            fontFamily: fontFamily,
            fontSize: fontSize,
            fontWeight: fontWeight,
            fontStyle: fontStyle
        ).resolvedFont(fallback: fallback)
    }
}

struct EditorMentionTheme {
    var textColor: UIColor?
    var backgroundColor: UIColor?
    var borderColor: UIColor?
    var borderWidth: CGFloat?
    var borderRadius: CGFloat?
    var fontWeight: String?
    var popoverBackgroundColor: UIColor?
    var popoverBorderColor: UIColor?
    var popoverBorderWidth: CGFloat?
    var popoverBorderRadius: CGFloat?
    var popoverShadowColor: UIColor?
    var optionTextColor: UIColor?
    var optionSecondaryTextColor: UIColor?
    var optionHighlightedBackgroundColor: UIColor?
    var optionHighlightedTextColor: UIColor?

    func merged(with override: EditorMentionTheme?) -> EditorMentionTheme {
        guard let override else { return self }
        var merged = self
        merged.textColor = override.textColor ?? merged.textColor
        merged.backgroundColor = override.backgroundColor ?? merged.backgroundColor
        merged.borderColor = override.borderColor ?? merged.borderColor
        merged.borderWidth = override.borderWidth ?? merged.borderWidth
        merged.borderRadius = override.borderRadius ?? merged.borderRadius
        merged.fontWeight = override.fontWeight ?? merged.fontWeight
        merged.popoverBackgroundColor =
            override.popoverBackgroundColor ?? merged.popoverBackgroundColor
        merged.popoverBorderColor = override.popoverBorderColor ?? merged.popoverBorderColor
        merged.popoverBorderWidth = override.popoverBorderWidth ?? merged.popoverBorderWidth
        merged.popoverBorderRadius = override.popoverBorderRadius ?? merged.popoverBorderRadius
        merged.popoverShadowColor = override.popoverShadowColor ?? merged.popoverShadowColor
        merged.optionTextColor = override.optionTextColor ?? merged.optionTextColor
        merged.optionSecondaryTextColor =
            override.optionSecondaryTextColor ?? merged.optionSecondaryTextColor
        merged.optionHighlightedBackgroundColor =
            override.optionHighlightedBackgroundColor ?? merged.optionHighlightedBackgroundColor
        merged.optionHighlightedTextColor =
            override.optionHighlightedTextColor ?? merged.optionHighlightedTextColor
        return merged
    }

    init(dictionary: [String: Any]) {
        textColor = EditorTheme.color(from: dictionary["textColor"])
        backgroundColor = EditorTheme.color(from: dictionary["backgroundColor"])
        borderColor = EditorTheme.color(from: dictionary["borderColor"])
        borderWidth = EditorTheme.cgFloat(dictionary["borderWidth"])
        borderRadius = EditorTheme.cgFloat(dictionary["borderRadius"])
        fontWeight = dictionary["fontWeight"] as? String
        popoverBackgroundColor = EditorTheme.color(from: dictionary["popoverBackgroundColor"])
        popoverBorderColor = EditorTheme.color(from: dictionary["popoverBorderColor"])
        popoverBorderWidth = EditorTheme.cgFloat(dictionary["popoverBorderWidth"])
        popoverBorderRadius = EditorTheme.cgFloat(dictionary["popoverBorderRadius"])
        popoverShadowColor = EditorTheme.color(from: dictionary["popoverShadowColor"])
        optionTextColor = EditorTheme.color(from: dictionary["optionTextColor"])
        optionSecondaryTextColor = EditorTheme.color(from: dictionary["optionSecondaryTextColor"])
        optionHighlightedBackgroundColor = EditorTheme.color(from: dictionary["optionHighlightedBackgroundColor"])
        optionHighlightedTextColor = EditorTheme.color(from: dictionary["optionHighlightedTextColor"])
    }
}

enum EditorToolbarAppearance: String {
    case custom
    case native
}

struct EditorToolbarTheme {
    var appearance: EditorToolbarAppearance?
    var height: CGFloat?
    var backgroundColor: UIColor?
    var borderColor: UIColor?
    var borderWidth: CGFloat?
    var borderRadius: CGFloat?
    var marginTop: CGFloat?
    var showTopBorder: Bool?
    var keyboardOffset: CGFloat?
    var horizontalInset: CGFloat?
    var separatorColor: UIColor?
    var buttonColor: UIColor?
    var buttonActiveColor: UIColor?
    var buttonDisabledColor: UIColor?
    var buttonActiveBackgroundColor: UIColor?
    var buttonBorderRadius: CGFloat?

    init(dictionary: [String: Any]) {
        appearance = (dictionary["appearance"] as? String).flatMap(EditorToolbarAppearance.init(rawValue:))
        height = EditorTheme.cgFloat(dictionary["height"])
        backgroundColor = EditorTheme.color(from: dictionary["backgroundColor"])
        borderColor = EditorTheme.color(from: dictionary["borderColor"])
        borderWidth = EditorTheme.cgFloat(dictionary["borderWidth"])
        borderRadius = EditorTheme.cgFloat(dictionary["borderRadius"])
        marginTop = EditorTheme.cgFloat(dictionary["marginTop"])
        showTopBorder = dictionary["showTopBorder"] as? Bool
        keyboardOffset = EditorTheme.cgFloat(dictionary["keyboardOffset"])
        horizontalInset = EditorTheme.cgFloat(dictionary["horizontalInset"])
        separatorColor = EditorTheme.color(from: dictionary["separatorColor"])
        buttonColor = EditorTheme.color(from: dictionary["buttonColor"])
        buttonActiveColor = EditorTheme.color(from: dictionary["buttonActiveColor"])
        buttonDisabledColor = EditorTheme.color(from: dictionary["buttonDisabledColor"])
        buttonActiveBackgroundColor = EditorTheme.color(from: dictionary["buttonActiveBackgroundColor"])
        buttonBorderRadius = EditorTheme.cgFloat(dictionary["buttonBorderRadius"])
    }

    var resolvedKeyboardOffset: CGFloat {
        keyboardOffset ?? (appearance == .native ? 6 : 0)
    }

    var resolvedHorizontalInset: CGFloat {
        horizontalInset ?? (appearance == .native ? 10 : 0)
    }

    var resolvedBorderRadius: CGFloat {
        borderRadius ?? (appearance == .native ? 20 : 0)
    }

    var resolvedBorderWidth: CGFloat {
        borderWidth ?? (appearance == .native ? 0 : 0.5)
    }

    var resolvedButtonBorderRadius: CGFloat {
        buttonBorderRadius ?? (appearance == .native ? 10 : 8)
    }
}

struct EditorContentInsets {
    var top: CGFloat?
    var right: CGFloat?
    var bottom: CGFloat?
    var left: CGFloat?

    init(dictionary: [String: Any]) {
        top = EditorTheme.cgFloat(dictionary["top"])
        right = EditorTheme.cgFloat(dictionary["right"])
        bottom = EditorTheme.cgFloat(dictionary["bottom"])
        left = EditorTheme.cgFloat(dictionary["left"])
    }
}

struct EditorTheme {
    var text: EditorTextStyle?
    var paragraph: EditorTextStyle?
    var blockquote: EditorBlockquoteTheme?
    var headings: [String: EditorTextStyle] = [:]
    var list: EditorListTheme?
    var horizontalRule: EditorHorizontalRuleTheme?
    var mentions: EditorMentionTheme?
    var links: EditorLinkTheme?
    var toolbar: EditorToolbarTheme?
    var placeholderColor: UIColor?
    var backgroundColor: UIColor?
    var borderRadius: CGFloat?
    var contentInsets: EditorContentInsets?

    static func from(json: String?) -> EditorTheme? {
        guard let json, !json.isEmpty,
              let data = json.data(using: .utf8),
              let raw = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return nil
        }
        return EditorTheme(dictionary: raw)
    }

    init(dictionary: [String: Any]) {
        if let text = dictionary["text"] as? [String: Any] {
            self.text = EditorTextStyle(dictionary: text)
        }
        if let paragraph = dictionary["paragraph"] as? [String: Any] {
            self.paragraph = EditorTextStyle(dictionary: paragraph)
        }
        if let blockquote = dictionary["blockquote"] as? [String: Any] {
            self.blockquote = EditorBlockquoteTheme(dictionary: blockquote)
        }
        if let headings = dictionary["headings"] as? [String: Any] {
            for level in ["h1", "h2", "h3", "h4", "h5", "h6"] {
                if let style = headings[level] as? [String: Any] {
                    self.headings[level] = EditorTextStyle(dictionary: style)
                }
            }
        }
        if let list = dictionary["list"] as? [String: Any] {
            self.list = EditorListTheme(dictionary: list)
        }
        if let horizontalRule = dictionary["horizontalRule"] as? [String: Any] {
            self.horizontalRule = EditorHorizontalRuleTheme(dictionary: horizontalRule)
        }
        if let mentions = dictionary["mentions"] as? [String: Any] {
            self.mentions = EditorMentionTheme(dictionary: mentions)
        }
        if let links = dictionary["links"] as? [String: Any] {
            self.links = EditorLinkTheme(dictionary: links)
        }
        if let toolbar = dictionary["toolbar"] as? [String: Any] {
            self.toolbar = EditorToolbarTheme(dictionary: toolbar)
        }
        placeholderColor = EditorTheme.color(from: dictionary["placeholderColor"])
        backgroundColor = EditorTheme.color(from: dictionary["backgroundColor"])
        borderRadius = EditorTheme.cgFloat(dictionary["borderRadius"])
        if let contentInsets = dictionary["contentInsets"] as? [String: Any] {
            self.contentInsets = EditorContentInsets(dictionary: contentInsets)
        }
    }

    func effectiveTextStyle(for nodeType: String, inBlockquote: Bool = false) -> EditorTextStyle {
        var style = text ?? EditorTextStyle()
        style = style.merged(with: inBlockquote ? blockquote?.text : nil)
        if nodeType == "paragraph" {
            style = style.merged(with: paragraph)
            if paragraph?.lineHeight == nil {
                style.lineHeight = nil
            }
        }
        style = style.merged(with: headings[nodeType])
        return style
    }

    static func cgFloat(_ value: Any?) -> CGFloat? {
        guard let number = value as? NSNumber else { return nil }
        return CGFloat(truncating: number)
    }

    static func fontWeight(from value: String) -> UIFont.Weight {
        switch value {
        case "100": return .ultraLight
        case "200": return .thin
        case "300": return .light
        case "500": return .medium
        case "600": return .semibold
        case "700", "bold": return .bold
        case "800": return .heavy
        case "900": return .black
        default: return .regular
        }
    }

    static func shouldApplyBoldTrait(_ value: String?) -> Bool {
        guard let value else { return false }
        return value == "bold" || Int(value).map { $0 >= 600 } == true
    }

    static func color(from value: Any?) -> UIColor? {
        guard let raw = value as? String else { return nil }
        let string = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()

        if let hexColor = colorFromHex(string) {
            return hexColor
        }
        if let rgbColor = colorFromRGBFunction(string) {
            return rgbColor
        }

        switch string {
        case "black": return .black
        case "white": return .white
        case "red": return .red
        case "green": return .green
        case "blue": return .blue
        case "gray", "grey": return .gray
        case "clear", "transparent": return .clear
        default: return nil
        }
    }

    private static func colorFromHex(_ string: String) -> UIColor? {
        guard string.hasPrefix("#") else { return nil }
        let hex = String(string.dropFirst())

        switch hex.count {
        case 3:
            let chars = Array(hex)
            return UIColor(
                red: component(String(repeating: String(chars[0]), count: 2)),
                green: component(String(repeating: String(chars[1]), count: 2)),
                blue: component(String(repeating: String(chars[2]), count: 2)),
                alpha: 1
            )
        case 4:
            let chars = Array(hex)
            return UIColor(
                red: component(String(repeating: String(chars[0]), count: 2)),
                green: component(String(repeating: String(chars[1]), count: 2)),
                blue: component(String(repeating: String(chars[2]), count: 2)),
                alpha: component(String(repeating: String(chars[3]), count: 2))
            )
        case 6:
            return UIColor(
                red: component(String(hex.prefix(2))),
                green: component(String(hex.dropFirst(2).prefix(2))),
                blue: component(String(hex.dropFirst(4).prefix(2))),
                alpha: 1
            )
        case 8:
            return UIColor(
                red: component(String(hex.prefix(2))),
                green: component(String(hex.dropFirst(2).prefix(2))),
                blue: component(String(hex.dropFirst(4).prefix(2))),
                alpha: component(String(hex.dropFirst(6).prefix(2)))
            )
        default:
            return nil
        }
    }

    private static func colorFromRGBFunction(_ string: String) -> UIColor? {
        let isRGBA = string.hasPrefix("rgba(") && string.hasSuffix(")")
        let isRGB = string.hasPrefix("rgb(") && string.hasSuffix(")")
        guard isRGBA || isRGB else { return nil }

        let start = string.index(string.startIndex, offsetBy: isRGBA ? 5 : 4)
        let end = string.index(before: string.endIndex)
        let parts = string[start..<end]
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }

        guard parts.count == (isRGBA ? 4 : 3),
              let red = Double(parts[0]),
              let green = Double(parts[1]),
              let blue = Double(parts[2])
        else {
            return nil
        }

        let alpha = isRGBA ? (Double(parts[3]) ?? 1) : 1
        return UIColor(
            red: red / 255,
            green: green / 255,
            blue: blue / 255,
            alpha: alpha
        )
    }

    private static func component(_ hex: String) -> CGFloat {
        CGFloat(Int(hex, radix: 16) ?? 0) / 255
    }
}
